include/tvm/ir/instrument.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/ir/instrument.h
+ *
+ * This file introduces a pass instrument infrastructure, inspired by LLVM and MLIR.
+ * It inserts instrumentation points around passes.
+ */
+#ifndef TVM_IR_INSTRUMENT_H_
+#define TVM_IR_INSTRUMENT_H_
+
+#include <tvm/node/reflection.h>
+#include <tvm/runtime/container.h>
+
+#include <utility>
+#include <vector>
+
+namespace tvm {
+
+class IRModule;
+
+// Forward class for PassInstrumentNode methods
+namespace transform {
+class PassInfo;
+}  // namespace transform
+
+namespace instrument {
+
+/*!
+ * \brief PassInstrumentNode forms an instrument implementation.
+ * It provides API for users to register callbacks at different instrumentation points.
+ *
+ * Within a pass context (tvm::transfom::PassContext), the instrumentation call sequence will like:
+ *
+ *   Instrument SetUp()
+ *
+ *     if (Instrument Before Pass())
```

Comment:
i think it would be great to separate the before-pass setup from the decision of "should this pass run?" The reason is that this removes concerns about how the ordering of pass instrumentation may change the number of calls to a given instrument. For example, if you just want to log all of the passes executed, you need to make sure that it's the first pass instrument registered.

I'd suggest this:
```
should_run_pass = [pi.should_run(pass) for pi in pass_instruments]
if all(should_run_pass):
  for pi in pass_instruments:
    pi.before_pass(pass)
  pass()
  for pi in pass_instruments:
    pi.after_pass(pass)
```

additionally, it would be great to document what happens if any of the non-pass logic raises an error here. the Python context manager framework should provide some good inspiration for this. it seems like it may be helpful to provide the after_pass with a "result" notion. also, can you add tests for these cases?

Reply:
Just checked LLVM and its logic is pretty like you commented.

Logic updated and test cases added.

For   _**providing the after_pass with a "result" notion**_, could you give more elaboration?

Reply:
hi @zackcquic, here i'm referencing the python `__exit__` function signature. you can see it used in context as "exit" [here](https://www.python.org/dev/peps/pep-0343/#specification-the-with-statement). i'm speaking fairly broadly, though--in Python, sys.exc_info() captures any pending exception as a tuple (type, value, traceback) whereas in c++ the most we get is information from catching e.g. `std::runtime_error`. By "result notation,"  I mainly mean some appropriate argument to `after_pass` to indicate why the `after_pass` is being invoked--it could be because:
1. a following `before_pass` errored
2. pass() was invoked and threw an error
3. pass() was invoked and finished cleanly
4. a preceding after_pass errored

Reply:
I add test cases and documentation for these cases' behavior in the newest commits. (include/tvm/ir/instrument.h, python/relay/tests_pass_instrument.py)

I assume if any error raise in `before_pass()`, `pass()` and `after_pass()`, it should be stopped, instead of keep processing.